test/python/pulse/test_builder.py

```diff
@@ -0,0 +1,969 @@
+        self.assertEqual(schedule, reference)
+
+    def test_inline(self):
+        """Test the inlining context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        with pulse.build() as schedule:
+            pulse.delay(3, d0)
+            with pulse.inline():
+                pulse.delay(5, d1)
+                pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference += instructions.Delay(5, d1)
+
+        self.assertEqual(schedule, reference)
+
+    def test_transpiler_settings(self):
+        """Test the transpiler settings context.
+
+        Tests that two cx gates are optimized away with higher optimization level.
+        """
+        twice_cx_qc = circuit.QuantumCircuit(2)
+        twice_cx_qc.cx(0, 1)
+        twice_cx_qc.cx(0, 1)
+
+        with pulse.build(self.backend) as schedule:
+            with pulse.transpiler_settings(optimization_level=0):
+                pulse.call_circuit(twice_cx_qc)
+        self.assertNotEqual(len(schedule.instructions), 0)
+
+        with pulse.build(self.backend) as schedule:
+            with pulse.transpiler_settings(optimization_level=3):
+                pulse.call_circuit(twice_cx_qc)
+        self.assertEqual(len(schedule.instructions), 0)
+
+    def test_scheduler_settings(self):
+        """Test the circuit scheduler settings context."""
+        inst_map = pulse.InstructionScheduleMap()
+        d0 = pulse.DriveChannel(0)
+        test_x_sched = pulse.Schedule()
+        test_x_sched += instructions.Delay(10, d0)
+        inst_map.add('x', (0,), test_x_sched)
+
+        x_qc = circuit.QuantumCircuit(2)
+        x_qc.x(0)
+
+        with pulse.build(backend=self.backend) as schedule:
+            with pulse.transpiler_settings(basis_gates=['x']):
+                with pulse.circuit_scheduler_settings(inst_map=inst_map):
+                    pulse.call_circuit(x_qc)
+
+        self.assertEqual(schedule, test_x_sched)
+
+    def test_phase_offset(self):
+        """Test the phase offset context."""
+        d0 = pulse.DriveChannel(0)
+
+        with pulse.build() as schedule:
+            with pulse.phase_offset(3.14, d0):
+                pulse.delay(10, d0)
+
+        reference = pulse.Schedule()
+        reference += instructions.ShiftPhase(3.14, d0)
+        reference += instructions.Delay(10, d0)
+        reference += instructions.ShiftPhase(-3.14, d0)
+
+        self.assertEqual(schedule, reference)
+
+    @unittest.expectedFailure
+    def test_frequency_offset(self):
+        """Test the frequency offset context.
+
+        .. note::
+            Expected to fail as ``ShiftFrequency`` is not yet implemented.
+        """
+        d0 = pulse.DriveChannel(0)
+
+        with pulse.build() as schedule:
+            with pulse.frequency_offset(1e9, d0):
```

Comment:
is `detuning` the word more commonly used for this?
```suggestion
            with pulse.detuning(1e9, d0):
```

Reply:
I agree detuning is more often used in an experimental context. My worry would be that if we end up with many instructions that have similar behavior but different names, ie., `phase_offset` has almost an identical meaning and implementation. Do you have thoughts @nkanazawa1989?

Reply:
bump